File: Code/Websites/DanpheEMR/wwwroot/DanpheApp/src/app/pharmacy/shared/phrm-invoice-return-items.model.ts

import {
  FormBuilder,
  FormControl,
  FormGroup,
  Validators
} from '@angular/forms';
export class PHRMInvoiceReturnItemsModel {
  public InvoiceReturnItemId: number = 0;
  public InvoiceReturnId: number = 0;
  public InvoiceItemId: number = null;
  public InvoiceId: number = null;
  public StoreId: number = 0;
  public BatchNo: string = "";
  public Quantity: number = 0;
  public TotalQty: number = 0;
  public AvailableQty: number = 0;//for UI only to view avl qty after  partially return
  public SalePrice: number = 0;
  public Price: number = 0;
  public SubTotal: number = 0;
  public VATPercentage: number = 0;
  public VATAmount: number = 0;
  public DiscountPercentage: number = 0;
  public DiscountAmount: number = 0;
  public TotalAmount: number = 0;
  public Remark: string = "";
  public CreatedBy: number = 0;
  public CreatedOn: string = "";
  public ReturnedQty: number = 0;
  public CounterId: number = 0;
  public checked: boolean;// for checking and uncheck item row in UI
  //some more variables for show only not for mapping with server
  public ItemId: number = 0;
  public ItemName: string = "";
  public SaledQty: number = 0;//all sale invoice qty details
  public FreeQty: number = 0;  //It's shows only FreeQty at sale time only for information
  public IsReturn: boolean = true;//it's flag for checkbox at client side for all select/deselect functionality

  public IsInsert: boolean = true;//mart Insert or update below is desc
  //When we post data to server
  //that data is PHRMInvoiceReturnItems table data,from this data some items already returned so , we need to update only quantity
  //and which records are return first time need to insert that record as new in PHRMInvoiceReturnItems

  public GRItemId: number = null;//only for refer update GRItem after sale return
  public ExpiryDate: string = null;
  public InvoiceItemsReturnValidator: FormGroup = null;
  // Added for manual return field
  public drugReturnItemObj: { ItemId: number, ItemName: string, GenericName: string };
  public availableBatches: { BatchNo: string; ExpiryDate: Date; SalePrice: number; }[] = [];
  public selectedBatch: { BatchNo: string; ExpiryDate: Date; SalePrice: number; } = null;
  public allowNewBatch: boolean = false;
  public GenericName: string; // for printing Generic Name in receipt
  public SoldQty: number; // for showing total main sold qty during invoice
  PreviouslyReturnedQty: number;
  TaxableAmount: number = 0;
  RackNo: string = '';
  PriceCategoryId: number = 0;
  //Constructor of class
  constructor() {
    var _formBuilder = new FormBuilder();
    this.InvoiceItemsReturnValidator = _formBuilder.group({
      'Quantity': ['', Validators.compose([this.positiveNumberValdiator])],
      'ReturnedQty': ['', Validators.compose([this.positiveNumberValdiator])]
      // 'Remark': ['', Validators.compose([Validators.required])],
    });
  }
  public IsDirty(fieldName): boolean {
    if (fieldName == undefined)
      return this.InvoiceItemsReturnValidator.dirty;
    else
      return this.InvoiceItemsReturnValidator.controls[fieldName].dirty;
  }

  public IsValid(): boolean { if (this.InvoiceItemsReturnValidator.valid) { return true; } else { return false; } } public IsValidCheck(fieldName, validator): boolean {
    if (fieldName == undefined)
      return this.InvoiceItemsReturnValidator.valid;
    else
      return !(this.InvoiceItemsReturnValidator.hasError(validator, fieldName));
  }

  positiveNumberValdiator(control: FormControl): { [key: string]: boolean } {
    if (control) {
      if (control.value <= 0)
        return { 'invalidNumber': true };
    }
  }

}

File: Code/Components/DanpheEMR.ServerModel/PharmacyModels/PHRMInvoiceReturnItemsModel.cs
using Audit.EntityFramework;
using System;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace DanpheEMR.ServerModel
{
    [AuditInclude]
    public class PHRMInvoiceReturnItemsModel
    {
        [Key]
        public int InvoiceReturnItemId { get; set; }
        public int InvoiceReturnId { get; set; }
        public int? InvoiceItemId { get; set; }
        public int? InvoiceId { get; set; }
        public int StoreId { get; set; }
        public string BatchNo { get; set; }
        public decimal Quantity { get; set; }
        public decimal SalePrice { get; set; }
        public decimal Price { get; set; }
        public decimal SubTotal { get; set; }
        public decimal VATPercentage { get; set; }
        public decimal DiscountPercentage { get; set; }
        public decimal TotalDisAmt { get; set; }
        public decimal PerItemDisAmt { get; set; }
        public decimal DiscountAmount { get; set; }
        public decimal TotalAmount { get; set; }
        public string Remark { get; set; }
        public int CreatedBy { get; set; }
        public int CounterId { get; set; }
        public DateTime CreatedOn { get; set; }
        public bool IsTransferredToACC { get; set; }
        public int ItemId { get; set; }
        public int CreditNoteNumber { get; set; }
        [NotMapped]
        public string FiscalYear { get; set; }
        public int FiscalYearId { get; set; }
        public decimal ReturnedQty { get; set; }
        [NotMapped]
        public bool IsInsert { get; set; }
        [NotMapped]
        public int? GRItemId { get; set; }

        [NotMapped]
        public string ItemName { get; set; }
        [NotMapped]
        public string GenericName { get; set; }
        [NotMapped]
        public decimal SaledQty { get; set; }
        [NotMapped]
        public decimal FreeQty { get; set; }
        public DateTime? ExpiryDate { get; set; }
        public PHRMInvoiceReturnItemsModel() => IsInsert = true;//expression block constructor      
        [NotMapped]
        public string ShortName { get; set; }
        [NotMapped]
        public string PANNumber { get; set; }
        [NotMapped]
        public string RackNo { get; set; }
        [NotMapped]
        public int InvoiceNo { get; set; }
        [NotMapped]
        public bool IsCoPayment { get; set; }
        [NotMapped]
        public decimal ReturnCashAmount { get; set; }
        [NotMapped]
        public decimal ReturnCreditAmount { get; set; }
        [NotMapped]
        public int? SchemeId { get; set; }
        [NotMapped]
        public int PatientId { get; set; }   
        [NotMapped]
        public int PriceCategoryId { get; set; }
    }
}
